File: ui/weather/daily_grid.py
from gi.repository import Gtk
from weather.base_grid import BaseGrid
from weather.temperature_icon import TemperatureIcon
from weather.temperature_label import TemperatureLabel


class DailyGrid(BaseGrid):
    def __init__(self):
        super().__init__(cols_count=4, rows_count=6)

        for row in range(0, self.rows_count):
            day = Gtk.Label.new("??")
            self.attach(day, 0, row, 1, 1)

            min_weather = TemperatureLabel()
            self.attach(min_weather, 1, row, 1, 1)

            max_weather = TemperatureLabel()
            self.attach(max_weather, 2, row, 1, 1)

            image = TemperatureIcon()
            self.attach(image, 3, row, 1, 1)

    def update_row(self, weather_on_day, row):
        self.get_child_at(0, row).set_label(weather_on_day.day)
        self.get_child_at(1, row).set_temperature(weather_on_day.temperature_min)
        self.get_child_at(2, row).set_temperature(weather_on_day.temperature_max)
        self.get_child_at(3, row).set_code(weather_on_day.code)

File: ui/src/weather/daily_grid.c
#include "ui/include/weather/daily_grid.h"
#include "ui/include/weather/base_grid.h"
#include "ui/include/weather/temperature_icon.h"
#include "ui/include/weather/temperature_label.h"

#define COLS_COUNT 4
#define ROWS_COUNT 6

void daily_grid_init(GtkWidget *self) {
  base_grid_init(self, COLS_COUNT, ROWS_COUNT);

  GtkWidget *child;
  for (size_t row = 0; row < ROWS_COUNT; row++) {
    child = gtk_label_new("??");
    gtk_grid_attach(GTK_GRID(self), child, 0, row, 1, 1);

    child = temperature_label_new();
    gtk_grid_attach(GTK_GRID(self), child, 1, row, 1, 1);

    child = temperature_label_new();
    gtk_grid_attach(GTK_GRID(self), child, 2, row, 1, 1);

    child = temperature_icon_new();
    gtk_grid_attach(GTK_GRID(self), child, 3, row, 1, 1);
  }
}

void daily_grid_refresh(GtkWidget *self, IO_CArray_WeatherOnDay weather) {
  GtkWidget *child;

  for (size_t row = 0; row < weather.len && row < ROWS_COUNT; row++) {
    IO_WeatherOnDay weather_on_day = weather.ptr[row];

    child = gtk_grid_get_child_at(GTK_GRID(self), 0, row);
    gtk_label_set_text(GTK_LABEL(child), weather_on_day.day);

    child = gtk_grid_get_child_at(GTK_GRID(self), 1, row);
    temperature_label_refresh(child, weather_on_day.temperature_min);

    child = gtk_grid_get_child_at(GTK_GRID(self), 2, row);
    temperature_label_refresh(child, weather_on_day.temperature_max);

    child = gtk_grid_get_child_at(GTK_GRID(self), 3, row);
    temperature_icon_refresh(child, weather_on_day.code);
  }
}
